core/shared/src/main/scala/fs2/internal/CompileScope.scala

```diff
@@ -346,30 +420,32 @@ private[fs2] final class CompileScope[F[_], O] private (
             }
         }
     }
+  }
 
   /**
     * When the stream is evaluated, there may be `Eval` that needs to be cancelled early,
     * when scope allows interruption.
     * Instead of just allowing eval to complete, this will race between eval and interruption promise.
-    * Then, if eval completes without interrupting, this will return on `Right`.
+    * Then, if eval completes without interrupting, this will return on `Right` - `Right`.
     *
-    * However when the evaluation is interrupted, then this evaluates on `Left`, where it either indicates
-    * the next stream step to take, when the evaluation was interrupted together with next scope.
-    * or the failure if the scope was interrupted by failure.
+    * However when the evaluation is normally interrupted the this evaluates on `Right` - `Left` where we signal
+    * what is the next scope from which we should calculate the next step to take.
     *
+    * Or if the evaluation is interrupted by a failure this evaluates on `Left` where the exception
+    * that caused the interruption is returned so that it can be handled.
     */
-  private[internal] def interruptibleEval[A](f: F[A]): F[Either[Throwable, Option[A]]] =
+  private[internal] def interruptibleEval[A](f: F[A]): F[Either[Throwable, Either[Token, A]]] =
```

Comment:
Perhaps we have to change this to F[Either[Either[Throwable, Token]], A] to prevent one allocation in case of normal evaluation.